src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -82,91 +88,63 @@ public void Dispose()
-                }
+                throw new InvalidOperationException(nameof(ReversedDiagnosticsServer.Start) + " method can only be called once.");
+            }
 
-                token.ThrowIfCancellationRequested();
+            _listenTask = ListenAsync(maxConnections, _disposalSource.Token);
+        }
 
-                if (null != stream)
-                {
-                    // Cancel parsing of advertise data after timeout period to
-                    // mitigate runtimes that write partial data and do not close the stream (avoid waiting forever).
-                    using var parseCancellationSource = new CancellationTokenSource();
-                    using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, parseCancellationSource.Token);
-                    try
-                    {
-                        parseCancellationSource.CancelAfter(ParseAdvertiseTimeout);
+        /// <summary>
+        /// Gets endpoint information when a new runtime instance connects to the server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>A task that completes with a <see cref="IpcEndpointInfo"/> value that contains information about the new runtime instance connection.</returns>
+        public async Task<IpcEndpointInfo> AcceptAsync(CancellationToken token)
+        {
+            VerifyNotDisposed();
 
-                        advertise = await IpcAdvertise.ParseAsync(stream, linkedSource.Token).ConfigureAwait(false);
-                    }
-                    catch (OperationCanceledException) when (parseCancellationSource.IsCancellationRequested)
-                    {
-                        // Only handle cancellation if it was due to the parse timeout.
-                    }
-                    catch (Exception ex) when (!(ex is OperationCanceledException))
-                    {
-                        // Catch all other exceptions and continue waiting for a new connection.
-                    }
-                }
+            VerifyIsStarted();
 
-                token.ThrowIfCancellationRequested();
+            var endpointInfoSource = new TaskCompletionSource<IpcEndpointInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
+            using var methodRegistration = token.Register(() => endpointInfoSource.TrySetCanceled(token));
+            using var disposalRegistration = _disposalSource.Token.Register(
+                () => endpointInfoSource.TrySetException(new ObjectDisposedException(nameof(ReversedDiagnosticsServer))));
 
-                if (null != advertise)
-                {
-                    Guid runtimeCookie = advertise.RuntimeInstanceCookie;
-                    int pid = unchecked((int)advertise.ProcessId);
+            RegisterEndpointInfoHandler((IpcEndpointInfo endpointInfo, out bool consumed) =>
```

Comment:
I'm fine if you want to put this in the future refactoring bucket but we've now got two copies of a very similar algorithm for doing producer/consumer operations over a list of items, one for streams and one for EndpointInfos. It appears we could de-dupe a bunch of code and probably make this more easily understood/testable if we factored out a type that looks like this:
````C#
class VisitableList<T> : IDisposable
{
    public delegate bool VisitFunc<T>(T item, out bool removeItem);

    // Adds an item to the list and invokes any visitors on it that are still processing items
    // VisitableList<Tuple<Guid,Stream>>.Add() replaces ProvideStream
    // VisitableList<EndpointInfo>.Add() replaces ProvideEndpointInfo
    public void Add(T item) 

    // runs the visitor over every current and future item in the list until the visit func
    // returns true or it is cancelled or disposed
    // whatever item the visitor was processing when it returned true is returned
    // if the removeItem returns true (and not cancelled/disposed) then the item is removed from the list
    public Task<T> VisitAsync(VisitFunc visit, CancellationToken token); 

    // sync version of VisitAsync
    public T Visit(VisitFunc visit, CancellationToken token);
    {
        VisitAsync(visit, token).Result;  // none of the typical async over sync deadlock risks should apply,
                                          // TaskCompletionSource returned by the Async func is a glorified ManualResetEvent
    }

    // The RegisterStreamHandler, RegisterEndpointInfoHandler should collapse as a private helper
}

// example rewritten APIs with this helper

        internal Stream Connect(Guid runtimeId, TimeSpan timeout)
        {
            VerifyNotDisposed();
            VerifyIsStarted();
            CancellationTokenSource cts = ...
            Tuple<Guid, Stream> result = _streamList.Visit((tuple, out removeItem) => 
            {
                removeItem = (tuple.First == runtimeId);
                return removeItem;
            }, cts.Token);
            return result.Second;
        }

        public async Task<IpcEndpointInfo> AcceptAsync(CancellationToken token)
        {
            VerifyNotDisposed();
            VerifyIsStarted();
            return await _endpointInfoList.VisitAsync((endpoint, out removeItem) => 
            {
                removeItem = true
                return true;
            }, token);
        }

       // The sync version of Visit() should make it very easy to create sync Accept()
````

Reply:
> I'm fine if you want to put this in the future refactoring bucket but we've now got two copies of a very similar algorithm for doing producer/consumer operations over a list of items, one for streams and one for EndpointInfos. It appears we could de-dupe a bunch of code and probably make this more easily understood/testable if we factored out a type that looks like this:

Yeah, I'm working on a similar refactor for the next change to add more of the non-async methods so that we have parity between non-async and async methods on the reversed server and related classes. This would have introduced a lot of code copying, and a big refactor, so I wanted to keep in separate from this change to separate the server loop from accepting new endpoints.